src/python/pants/backend/go/util_rules/go_mod.py

```diff
@@ -58,14 +102,125 @@ async def find_nearest_go_mod(request: OwningGoModRequest) -> OwningGoMod:
+            f"The `{alias}` target `{address}` requires that the address of an owning `{GoModTarget.alias}` "
+            f"target be given via the `{GoOwningGoModAddressField.alias}` field since it is "
+            "a dependency of a Go target. However, there are multiple `go_mod` targets in this repository "
+            "which makes the choice of which `go_mod` target to use ambiguous. Please specify which of the "
+            "following addresses to use or consider using the `parametrize` builtin to specify more than "
+            "one of these addresses if this target will be used in multiple Go modules: "
+            f"{', '.join([str(tgt.address) for tgt in all_go_mod_targets])}"
+        )
+
+    # If a value is specified, then resolve it as an address.
+    address_input = AddressInput.parse(
+        field.value,
+        relative_to=address.spec_path,
+        description_of_origin=f"the `{GoOwningGoModAddressField.alias}` field of target `{address}`",
+    )
+    candidate_go_mod_address = await Get(Address, AddressInput, address_input)
+    wrapped_target = await Get(
+        WrappedTarget,
+        WrappedTargetRequest(
+            candidate_go_mod_address,
+            description_of_origin=f"the `{GoOwningGoModAddressField.alias}` field of target `{address}`",
+        ),
+    )
+    if not wrapped_target.target.has_field(GoModDependenciesField):
+        raise InvalidTargetException(
+            f"The `{alias}` target `{address}` requires that the address of an owning `{GoModTarget.alias}` "
+            f"target be given via the `{GoOwningGoModAddressField.alias}` field since it is "
+            f"a dependency of a Go target. However, the provided address `{field.value}` does not refer to "
+            f"a `{GoModTarget.alias}` target. Please specify which of the following addresses to use or consider "
+            "using the `parametrize` builtin to specify more than one of these addresses if this target will be "
+            f"used in multiple Go modules: {', '.join([str(tgt.address) for tgt in all_go_mod_targets])}"
+        )
+    return candidate_go_mod_address
+
+
+@rule
+async def find_owning_go_mod(
+    request: OwningGoModRequest, all_go_mod_targets: AllGoModTargets
+) -> OwningGoMod:
+    wrapped_target = await Get(
+        WrappedTarget,
+        WrappedTargetRequest(request.address, description_of_origin="the `OwningGoMod` rule"),
+    )
+    target = wrapped_target.target
+
+    if target.has_field(GoModDependenciesField):
+        return OwningGoMod(request.address)
+    elif (
+        target.has_field(GoPackageSourcesField)
+        or target.has_field(FileSourceField)
+        or target.has_field(FilesGeneratingSourcesField)
+        or target.has_field(ResourceSourceField)
+        or target.has_field(ResourcesGeneratingSourcesField)
+    ):
+        # For `go_package`, `file`, and `resource` targets, use the nearest ancestor `go_mod` target.
+        nearest_go_mod_result = await Get(
```

Comment:
@Eric-Arellano: Maybe we should always check for nearest ancestor `go_mod` before checking what kind of target is being examined?